### PR TITLE
MonadError instance for Kleisli

### LINKED ISSUE #1542: MonadError instance for Kleisli
Is there any reason for not having a `MonadError` instance for `Kleisli`? Since `MonadError` seems to be simply `Monad` and `ApplicativeError`, and `Kleisli` has instances of both, it seems like it could work. Although prioritizing the instance could be a little tricky, I guess...

### PR DESCRIPTION
**Pull request by durban**
Fixes #1542.

### CODE REVIEW

**Comment by kailuowang** [comment]
`MonadError` is more specific than `ApplicativeError` and thus should be placed in higher priority, no?

**Comment by durban** [comment]
I'm not aware of such a rule (and if there is one, it is already broken by `Applicative` and `MonadReader` for `Kleisli`). Is this documented somewhere? Can it cause any problems having these with a lower priority?

**Comment by kailuowang** [comment]
@durban, you are right, I believe `MonadReader` is in the wrong priority according to this rule (mind to fix this on the way as well?). I don't think this rule is documented anywhere here, it's not a cats specific rule, it's more a best practice to my knowledge. My understanding is that this pratice has two benefits: 
1) having an implicit sequence is better not having one.
2) if you assign the most specific one with the highest priority, it takes less implicit search, so it's slightly more compilation efficient.  And sometimes the most specific (and thus powerful) instance is also more performant.
I might've missed other benefits.

File: core/src/main/scala/cats/data/Kleisli.scala [hunk 1]
```diff
@@ -275,6 +279,20 @@ private trait KleisliApplicativeError[F[_], A, E] extends KleisliApplicative[F,
...
+
+  def flatMap[B, C](fa: Kleisli[F, A, B])(f: B => Kleisli[F, A, C]): Kleisli[F, A, C] =
+    fa.flatMap(f)
+
+  def map[B, C](fa: Kleisli[F, A, B])(f: B => C): Kleisli[F, A, C] =
+    fa.map(f)
+
+  def tailRecM[B, C](b: B)(f: B => Kleisli[F, A, Either[B, C]]): Kleisli[F, A, C] =
+    Kleisli[F, A, C]({ a => FlatMap[F].tailRecM(b) { f(_).run(a) } })
+}
```
Comment (peterneyens) [inline core/src/main/scala/cats/data/Kleisli.scala]: Like this we have the some code twice:

- here in `KleisliFlatMap`
- and in `catsDataMonadReaderForKleisli`

Maybe we could create a `KleisliMonad` which can be extended by a `KleisliMonadReader` and a `KleisliMonadError` (similar as how it is done in `StateT`) ?

**Comment by durban** [comment]
@kailuowang @peterneyens I've amended my commit to address your comments. (I've also rebased to the current master.)

**Pull request merged**